### PR TITLE
new GetGCMemoryInfo API

### PR DESCRIPTION
**Pull request by Maoni0**
Most of the changes inside GC are the following -

All the info is updated at the end of a GC with the exception of BGC which needs special care. We maintain 2 BGC info
data structures and one could be in the process of getting updated and if the user requests either BGC info or the last
GC info which is also a BGC, we would return the one that's complete. See comments in code for a more detailed explanation.

Fill in the after size data for all generations instead of just the affected generations.

Timing info recording -

+ Changed the GC time recording from ms to us. Previously we did the timing analysis with events which also carry their
own high res timestamps so in GC using ms is good enough. But now since we are returning the timing info we want to
record it in us instead of ms.
+ Consolidated when we get end_gc_time. It was not including all of the GC work before (eg, decommit eph seg). And moving
it to do_post_gc makes it easier for getting the GC info.
+ Not including the EE restart time for blocking GCs - the reason being restart takes very little time and by the time we
get to the actual RestartEE the GC settings could be changed and it's not worth the effort to wait till RestartEE at which
point we have to spend more effort to figure out which GC to attribute it to.
+ Maintain the running % Pause time in GC info.

### CODE REVIEW

**Comment by stephentoub** [comment]
Did you accidentally miss including a file containing GCMemoryInfoData?

**Comment by Maoni0** [comment]
I see it in src/coreclr/src/System.Private.CoreLib/src/System/GC.cs, is that not what you are seeing?

**Comment by jkotas** [comment]
Large number of build legs is failing with:
```
src/libraries/System.Private.CoreLib/src/System/GCMemoryInfo.cs(20,26): error CS0246: The type or namespace name 'GCMemoryInfoData' could not be found (are you missing a using directive or an assembly reference?)
```
After closer look, these build errors are in Mono build. So you will need to do something to make the Mono build green.

**Comment by PeterSolMS** [comment]
I wonder about this:

size_t GetHighPrecisionTimeStamp()
{
    int64_t ts = GCToOSInterface::QueryPerformanceCounter();

    return (size_t)(ts / qpf_us);
}

qpf_us being an int64_t, there could be a round-off error on some systems. It's probably more precise and efficient to use a double qpf_interval_us = 1e6 / qpf and do return (size_t)(ts * qpf_interval) instead.

**Comment by PeterSolMS** [comment]
I think the doc comment here needs a few words more:

        /// <summary>
        /// The index of this GC.
        /// </summary>
        public long Index => _data._index;

suggestion:

        /// The index of this GC. Indices start at 1 and increase by 1 for each GC, so the third GC in the execution of the program will have index 3.

But perhaps you can think of something better...

**Comment by Maoni0** [comment]
@jkotas right, it was just mono build failures; I have since moved the definition of `GCMemoryInfoData` from src\coreclr\src\System.Private.CoreLib\src\SystemGC.cs to src\libraries\System.Private.CoreLib\src\System\GCMemoryInfo.cs, hopefully that addresses the issue. too bad this kind of failures do not show up when we build the coreclr version.

@PeterSolMS I've added details to the comment of Index and also changed the type of `qpf_us` to double and the return type of `GetHighPrecisionTimeStamp` to `uint64_t`.

**Comment by Maoni0** [comment]
@marek-safar, what's the process of making it work for mono? obviously mono's implementation is not there yet.

**Comment by marek-safar** [comment]
@Maoni0 for this case I'd suggest to

- Create a new issue for Mono'GC to implement the support
- Link the issue when disabling newly added tests for MonoVM
- Add public API to Mono's SPC and make it throw PNSE (would be nice to have the enums in shared part)

**Comment by Maoni0** [comment]
thanks @marek-safar!

> Create a new issue for Mono'GC to implement the support

https://github.com/dotnet/runtime/issues/37950
    
> Link the issue when disabling newly added tests for MonoVM

I didn't add tests to MonoVM. I added tests in coreclr. I'd like this test to run standalone. as I understand the tests in libs can be run in a process with other tests.

> Add public API to Mono's SPC and make it throw PNSE (would be nice to have the enums in shared part)

I have no idea what SPC or PNSE means :) could you please elaborate, or better yet, do you have an example? I'll move the enums in the shared part.

**Comment by PeterSolMS** [comment]
I wonder what happens if there was no GC of the requested kind yet - reading the code, it looks like all the fields will contain zeroes, in particular the index, which would be non-zero normally. 

If my reading of the code is correct, should we document that the reliable way to detect that no GC of the requested has happened yet is to check if the index is 0?

**Comment by Maoni0** [comment]
if there's no GC of the kind that happened, then by design you'd get back all 0's including the index. we can mention this in the doc.

**Comment by marek-safar** [comment]
> I didn't add tests to MonoVM. I added tests in coreclr. I'd like this test to run standalone.

These tests are run also with MonoVM so expect failures there. I know the location is wrong, the naming is wrong but that's being fixed.

> I have no idea what SPC or PNSE means :) could you please elaborate

As you are adding public API into CoreCLR specific implementation you need to add same public API to Mono's specific System.Private.CoreLib -- https://github.com/dotnet/runtime/blob/master/src/mono/netcore/System.Private.CoreLib/src/System/GC.Mono.cs. It should be enough to add them with throw PlatformNotSupportedException (PNSE) for now.

**Comment by PeterSolMS** [comment]
It looks like in this computation ephemeral_elapsed is now in microseconds - need to divide by 1000 to get milliseconds:

size_t ephemeral_elapsed = (size_t)(dd_time_clock (dd0) - gc_last_ephemeral_decommit_time)

**Comment by PeterSolMS** [comment]
I think you implicitly changed the semantics of this method to return microseconds instead of milliseconds - probably unintentional?

size_t GCHeap::GetLastGCDuration(int generation)
{
#ifdef MULTIPLE_HEAPS
    gc_heap* hp = gc_heap::g_heaps[0];
#else
    gc_heap* hp = pGenGCHeap;
#endif //MULTIPLE_HEAPS
    return dd_gc_elapsed_time (hp->dynamic_data_of (generation));
}

**Review by PeterSolMS — APPROVED** [review]

**Comment by Maoni0** [comment]
@marek-safar got it, thanks so much!

**Comment by Maoni0** [comment]
@PeterSolMS thanks so much for finding those!!! I'll make the corresponding changes a bit later today.

**Comment by jkotas** [comment]
> documenting the new GetGCMemoryInfo API

You should just include the docs under triple-slash comments in the doc. The doc generation tool will use to populate the docs for newly added APIs.

**Comment by Maoni0** [comment]
okie; I didn't know that was all there is for API doc... I thought we'd have sections that you don't see as code comments like "examples of using this API" and etc.

**Comment by stephentoub** [comment]
> I thought we'd have sections that you don't see as code comments like "examples of using this API" and etc.

We (and others) can always submit PRs to the docs repo to augment the content further after they're initially populated.  We just use the doc comments to seed the minimum so we at least have basic docs for summary, parameters, etc.

File: src/coreclr/src/System.Private.CoreLib/src/System/GC.cs
Comment (stephentoub) [inline src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:64]: Can this just call the other overload?
```C#
public static GCMemoryInfo GetGCMemoryInfo() => GetGCMemoryInfo((GCKind)(-1));
```

File: src/coreclr/src/System.Private.CoreLib/src/System/GC.cs
Comment (stephentoub) [inline src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:66]: Should this do any validation on GCKind to make sure it's one of the expected values?

File: src/coreclr/src/System.Private.CoreLib/src/System/GC.cs
Comment (stephentoub) [inline src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:69]: -1 ends up being valid.  Should it be in the enum?

File: src/libraries/System.Private.CoreLib/src/System/GCMemoryInfo.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/GCMemoryInfo.cs:12]: This should get triple-slash doc comment as well to get auto-generated documentation

File: src/libraries/System.Private.CoreLib/src/System/GCMemoryInfo.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/GCMemoryInfo.cs:23]: Dtto

File: src/mono/netcore/System.Private.CoreLib/src/System/GC.Mono.cs
Comment (stephentoub) [inline src/mono/netcore/System.Private.CoreLib/src/System/GC.Mono.cs:251]: It looks like this overload used to work, but it's now being changed to throw?

**Comment by GSPP** [comment]
> if there's no GC of the kind that happened, then by design you'd get back all 0's including the index. we can mention this in the doc.

That's nasty... Maybe return a nullable value instead? Or `bool` return and `out` parameter?

File: src/coreclr/src/System.Private.CoreLib/src/System/GC.cs
Comment (Maoni0) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:64]: it could :) I'll change it.

File: src/coreclr/src/System.Private.CoreLib/src/System/GC.cs
Comment (Maoni0) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:66]: I do have an assert in the GC code but I agree, there should be one on the BCL side.

File: src/coreclr/src/System.Private.CoreLib/src/System/GC.cs
Comment (Maoni0) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:69]: I thought about making it part of the enum but since this was the API surface we agreed on I didn't change it. should I make it part of the enum? I'm fine either way.

File: src/libraries/System.Private.CoreLib/src/System/GCMemoryInfo.cs
Comment (Maoni0) [reply src/libraries/System.Private.CoreLib/src/System/GCMemoryInfo.cs:12]: ahh right, will do.

File: src/mono/netcore/System.Private.CoreLib/src/System/GC.Mono.cs
Comment (Maoni0) [reply src/mono/netcore/System.Private.CoreLib/src/System/GC.Mono.cs:251]: well, because there isn't a mono implementation yet. I did this based on @marek-safar's suggestion but I dunno if I misunderstood him.

File: src/coreclr/src/System.Private.CoreLib/src/System/GC.cs
Comment (stephentoub) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:69]: What does -1 end up meaning... "All"?

Comment (Maoni0) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:69]: -1 means any. change to this? 
```csharp
    public enum GCKind
    {
        Ephemeral = 0,    // gen0 or gen1 GC
        FullBlocking = 1, // blocking gen2 GC
        Background = 2, // background GC (always gen2)
        Any = 3               // any kind
    };
```

Comment (stephentoub) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:69]: I do think we should add it to the enum, and Any sounds like a fine name, though I'm not sure about the numerical value.

@terrajobst, @bartonjs, opinions?

Comment (Maoni0) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:69]: just to clarify - Ephmeral/FullBlocking/Background are all exclusive - a GC can only be one of these 3 kinds. and Any means any **one** of these 3 kinds.

Comment (stephentoub) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:69]: Understood.  My question about the value is when in the future we introduce a new kind and it gets value 4, it's a little strange that any is in the middle.

Comment (Maoni0) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:69]: I see; it could be a much larger number like 0xff?

Comment (jkotas) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:69]: `0` would be the most natural choice.

Comment (Maoni0) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:69]: 0 SGTM.

**Comment by Maoni0** [comment]
I believe I've addressed all your concerns, @jkotas and @stephentoub, please take a look at the latest commit. the only thing I'm waiting on is the review from mono folks. @BrzVlad I have no experience doing this; just let me know if I did something wrong.

File: src/coreclr/src/System.Private.CoreLib/src/System/GC.cs
Comment (jkotas) [inline src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:63]: The usual style for error handling is to put it first, e.g.:

```
if ((kind <  GCKind.Any) || (kind > GCKind.Background))
    throw new ...

var data = new GCMemoryInfoData();
GetMemoryInfo(data, (int)kind);
return new GCMemoryInfo(data);
```

Comment (Maoni0) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs]: fixed

File: src/mono/netcore/System.Private.CoreLib/src/System/GC.Mono.cs
Comment (BrzVlad) [reply src/mono/netcore/System.Private.CoreLib/src/System/GC.Mono.cs:251]: @Maoni0 I think @marek-safar's suggestion originated from the assumption that this is a new API that we don't yet support and we should just throw exception for now. However, this API turned out to be an extension on the existing `GCMemoryInfo`, and, as Stephen pointed out, we already support part of it and I don't think we should regress what is already implemented. 

We used to just leave fields that we don't yet support uninitialized in the GCMemoryInfo (we might not even support every single field right now) and I remember coreclr was doing the same at some point, when comparing between Windows and Linux, for example. My opinion is that we should leave the fields from the new API with the default value for now and just initialize the fields from the netcore 3.0 version. So instead of throwing, we could just create a new `GCMemoryInfoData` with just these 5 fields initialized and we'll add later support in the mono runtime for the rest of the api.

Comment (Maoni0) [reply src/mono/netcore/System.Private.CoreLib/src/System/GC.Mono.cs]: I see; makes sense! does [this](https://github.com/dotnet/runtime/pull/37879/commits/27c06c8b822a07993f07af15c291f21fa1273f16) look good to you? I'm throwing PNSE in the overload but changed the old API the way you suggested.

**Comment by Maoni0** [comment]
this is good to go; I will wait till tomorrow noon and if I don't hear any objections by then I will merge. 

thank you all for your reviews; really appreciate your help!

**Review by BrzVlad — APPROVED** [review]

**Pull request merged**
